### ISSUE TITLE
Glitch when updating a Var while reading its state

### ISSUE THREAD

**Issue by raquo** [raquo/Airstream#39]
While Airstream solves vast majority of FRP glitches, a particular grief-causing pattern has emerged:

```scala
var n = 0
val clickBus = new EventBus[Unit]
val state = Var[List[Int]](Nil)
clickBus
  .events
  .map(_ => n = n + 2)
  .flatMap(_ => EventStream.seq(List(n -2, n - 1), emitOnce = true))
  .withCurrentValueOf(state.signal)
  .foreach { (next, currState) =>
    state.set(currState :+ next)
  }
}
clickBus.writer.onNext(())
println(state.now())
```

source: [gitter](https://gitter.im/Laminar_/Lobby?at=5f7239b75a56b467a5f7094c)

User expects state to contain `List(0, 1)` at the end, however in fact `state` will contain only `List(1)` because when the second event happens, `currState` will still be `Nil` because every `state.set` call will only update Var's state in a new transaction, which is created and thus scheduled AFTER both EventStream.seq transactions were created & scheduled, so it will also run AFTER they have completed.

To simplify, here is a similar pattern that results in the same issue for similar reasons:

```scala
var n = 0
val clickBus = new EventBus[Unit]
val state = Var[List[Int]](Nil)
clickBus
  .events
  .foreach { _ =>
    n = n + 2
    state.set(state.now() :+ (n - 2)
    state.set(state.now() :+ (n - 1)
  }
}
clickBus.writer.onNext(())
println(state.now())
```

source: also gitter, but older, can't find it now

In the example above, the call to `state.now()` will return `Nil` both times, because the transaction kicked off by `state.set(state.now() :+ (n - 2)` hasn't processed yet, because the current transaction hasn't finished yet.

One solution could be to replace `state.set(state.now() :+ (n - 1)` with `new Transaction(_ => state.set(state.now() :+ (n - 1))`, which would delay evaluation of `state.now()` until a new transaction which is scheduled after the transaction created by `state.set(state.now() :+ (n - 2)`. However, this is ugly both visually and conceptually. This call will actually create two transactions. Maybe we can add a `nowSync(currValue => ())` to Var which would trigger the callback only after the current transaction or even ALL transactions have been cleared?

Alternatively, we already have `update(mod: A => A)` method on Var, however it evaluates the next state immediately when called. Perhaps it should instead delay the evaluation until the transaction that it creates actually starts executing.

Currently we have:

```
def update(mod: A => A): Unit = {
    val unsafeValue = now()
    val nextValue = Try(mod(unsafeValue))
    new Transaction(setCurrentValue(nextValue, _))
  }
```

It would be pretty simple to move `unsafeValue` and `nextValue` evaluation inside the Transaction callback.

Yet another alternative is for the user to introduce an asynchronous delay somewhere between two events that they fire. So add `delay(0)` to `EventStream.seq(List(n -2, n - 1), emitOnce = true)` in the first example, or wrap `state.set(state.now() :+ (n - 1)` in `js.timers.setTimeout` in the second one.

Also I should document this issue and workarounds more clearly in Laminar & Airstream docs.

If anyone has other ideas or concerns about proposed solutions, please speak up. This issue will likely require breaking changes to the API where it will still compile but the timing logic would be different, requiring manual review of your usages of Var.update and such.

**Comment by ngbinh** [comment]
Thanks for the extensive documentation!

**Comment by raquo** [comment]
Another solution by Iurii for certain cases – [use `fold`](https://gitter.im/Laminar_/Lobby?at=5f72953b8fe6f119636a0c78) as a way to read previous state without introducing a transaction boundary with Var.

**Comment by raquo** [comment]
Just wanted to add some more explanation while I remember:

In the very first snippet in this issue, here is exactly why the first `state.set` call happens AFTER the transaction for the second emitted number was scheduled:

1. clickBus transaction starts
2. We create EventSream.seq which creates two transactions
3. But because clickBus transaction is not finished yet, we schedule both of those transactions instead of running the first one
4. clickBus propagation ends and the next transaction, for the first number, starts running
5. we reach `state.set` for the first time, but the transaction for the second event has already been scheduled, so we can only schedule the new transaction for after it

This suggests an alternative approach: change the transaction scheduling logic inside `EventStream.fromSeq` (and more importantly, inside `EventStream.merge`, which works similarly) to only schedule the next transaction after the previous transaction was done.

This will solve our timing with `state.set`, however I'm not sure if it's a good solution. For example, if the result of state.set causes yet another transaction to be started, e.g. by means of another `flatMap` or `anotherState.set`, those updates would not be scheduled until AFTER the second EventStream.seq number event transaction. This would cause a sort of weird situation where this second number transaction is scheduled at what might feel like an arbitrary point in time in some cases.

Instead, maybe the timing logic inside `EventStream.seq` and `EventStream.merge` should be such that the second event is scheduled (and fired) only after the first transaction has completed AND after the pending transactions queue has cleared after that. Meaning, if running the first transaction caused more transactions to happen, those would be processed first, and only after all that is done, would the next transaction fire.

But, what about other behaviours? You can create an EventBus and fire two events into it – if you're not doing this inside a transaction already, it will execute the first event transaction and all of the transactions it creates before proceeding to the next line of code that fires another event, which would do the same before proceeding to the next line of code.

However, if on the other hand we're inside a transaction already, the timing logic will be different – both event transactions will be scheduled for the future instead of executed immediately. I think maybe I finally see the general issue here: when a transaction is created from within the currently running transaction, it needs to be scheduled _after_ all transactions previously scheduled _from this exact current transaction_ but _before_ all _other_ transactions in the queue. What would those other transactions be? Those scheduled from within previous "current transactions", i.e. ancestors of the currently running transaction.

So... you know how we have a depth-first propagation strategy? Within a given transaction, we propagate events from current observable to its first child, and to that observable's first child, and so on, until we exhaust that line and back up one level. But on the transactions layer we're sort of doing the opposite, we're doing it _sort of_ breadth-first, i.e. if you call state.set twice, we'll schedule both of these immediately and then if the first state.set transaction results in more child transactions, we will not execute them first, but rather back up and run the second state.set transaction first. I wonder if running those other transactions first as a general design principle of Airstream would work better. Would it work perfectly? What kind of glitches would it solve? What kind of new glitches, if any, would it create? What kind of data structure would we need for the transactions queue now? If it's simpler than a tree, can we use the same structure for observables' dependencies? Why and why not?

This would be a pretty big breaking change that, assuming it solves the problem, would be absolutely worth it, but it really is a hard problem, and I will need to be thinking about it for a while.

**Comment by raquo** [comment]
Another variation of the same:

```scala
    val bus = new EventBus[Int]
    val log = Var[List[Int]](Nil)
    val stream1 = bus.events
    val stream2 = bus.events.map(_ * 100)
    val obs = Observer[Int] { num =>
      log.update(_:+ num)
    }
    mount(
      div(
        stream1 --> obs,
        stream2 --> obs
      )
    )

    bus.writer.onNext(1)

    log.now() shouldBe List(1, 100) // Fails, is actually List(100)
```

Workaround the same: add delay(0) to `stream1` or `stream2`

**Comment by raquo** [comment]
Anyone affected by this, please try the new release, see #40.

**Comment by vic** [comment]
Fixed by #40
